### PR TITLE
Prevent self nesting

### LINKED ISSUE #1944: Prevent self-nested variables
Creates an infinite loop:
```.py
v = sc.empty(dims=[], shape=[], dtype=sc.dtype.Variable)
v.value = v
print(v)
```
Also happens with `dtype=sc.dtype.PyObject` which might be harder to detect. But at least that one makes Python raise RecursionError.

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #1944

Only covers cases where variables, DataArrays, or Datasets are directly assigned to each other. It is still possible to create reference loops using more deeply nested constructs like
```.cpp
Variable v = makeVariable<DataArray>(Values{DataArray{makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{1, 2})}});  
v.value<DataArray>().attrs().set(Dim::X, v);
```
Catching this requires a check involving `Dict` and potentially a pointer to its parent.

### CODE REVIEW

File: dataset/include/scipp/dataset/dataset_util.h [hunk 1]
```diff
@@ -8,15 +8,67 @@
...
+void check_nested_in_assign(const T &lhs, const Variable &rhs) {
+  if (rhs.dtype() == dtype<DataArray>) {
+    for (const auto &nested : rhs.values<DataArray>()) {
+      check_nested_in_assign(lhs, nested);
+    }
+  } else if (rhs.dtype() == dtype<Dataset>) {
+    for (const auto &nested : rhs.values<Dataset>()) {
+      check_nested_in_assign(lhs, nested);
+    }
+  }
```
Comment (SimonHeybrock) [inline dataset/include/scipp/dataset/dataset_util.h:36]: Do we also need to check the buffer of a binned variable here?

File: dataset/include/scipp/dataset/dataset_util.h [hunk 2]
```diff
@@ -8,15 +8,67 @@
...
+void check_nested_in_assign(const T &lhs, const Dict<Key, Value> &rhs) {
+  for (const auto &[_, var] : rhs) {
+    check_nested_in_assign(lhs, var);
+  }
+}
+
+template <class L, class R>
+void check_nested_in_assign(const L &lhs, const R &rhs) {
+  if constexpr (std::is_same_v<L, R>) {
+    if (&lhs == &rhs) {
```
Comment (SimonHeybrock) [inline dataset/include/scipp/dataset/dataset_util.h:49]: Is this the endpoint of the recursive `check_nested_in_assign` calls? If so, it is not 100% clear to me (at the end of a workday) why pointer comparison of (e.g., DataArray?) is sufficient here. Wouldn't we also need to compare buffer pointers (`VariableConceptHandle`)? Or is that what is happening?

File: variable/variable.cpp [hunk 3]
```diff
@@ -29,6 +27,45 @@ Variable::Variable(const Dimensions &dims, VariableConceptHandle data)
...
+} // namespace
+
+Variable &Variable::operator=(const Variable &other) {
+  check_nested_in_assign(*this, other);
+  m_dims = other.m_dims;
+  m_strides = other.m_strides;
+  m_offset = other.m_offset;
+  m_object = other.m_object;
+  m_readonly = other.m_readonly;
+  return *this;
```
Comment (SimonHeybrock) [inline variable/variable.cpp:56]:
```suggestion
  return *this = Variable(other);
```
?

File: dataset/include/scipp/dataset/dataset_util.h (hunk 1 above)
Comment (jl-wynen) [reply dataset/include/scipp/dataset/dataset_util.h:36]: Correct me if I'm wrong but Variables of Variables cannot be binned. So the only case where we would need to check the buffer is if that is a DataArray. The only way to get a cycle in this case would be trough
```.py
da = get_binned_data_array()
var = sc.scalar(da)
da.attrs['attr'] = var
```
which requires a check in `Dict` and not `Variable`.

File: dataset/include/scipp/dataset/dataset_util.h (hunk 2 above)
Comment (jl-wynen) [reply dataset/include/scipp/dataset/dataset_util.h:49]: As far as I understand it, storing DataArrays or Datasets in buffers copies them. So we can only get a cycle through constructs like the one in the PR description which are much harder to detect.

File: variable/variable.cpp (hunk 3 above)
Comment (jl-wynen) [reply variable/variable.cpp:56]: Possible. Do you trust the optimizer enough?

Comment (SimonHeybrock) [reply variable/variable.cpp:56]: Why optimizer? It is just using move semantics, which are guarenteed?

Comment (jl-wynen) [reply variable/variable.cpp:56]: Sure. But it is a copy + move instead of just a copy.

Comment (SimonHeybrock) [reply variable/variable.cpp:56]: True. Should be cheap though (compared to typical variable sizes)? And as you say, to optimizer might take care of it.

File: dataset/test/data_array_test.cpp [hunk 4]
```diff
@@ -137,3 +137,30 @@ TEST_F(DataArrayTest, full_slice) {
...
+
+  DataArray nested_in_data{var};
+  ASSERT_THROW_DISCARD(var.value<DataArray>() = nested_in_data,
+                       std::invalid_argument);
+
+  DataArray nested_in_coord{
+      makeVariable<double>(Dims{Dim::X}, Shape{2}, Values{3, 4})};
+  nested_in_coord.coords().set(Dim::X, var);
+  ASSERT_THROW_DISCARD(var.value<DataArray>() = nested_in_coord,
+                       std::invalid_argument);
```
Comment (SimonHeybrock) [inline dataset/test/data_array_test.cpp:153]: So if I understand correctly this and the ones below detects cases such as

```python
da.coords['abc'].value = da
```

What about setting up the same, but in reverse order:

```python
da.coords['abc'] = sc.scalar(da)
```

which I believe is much more likely to occur in practice.

File: dataset/include/scipp/dataset/dataset_util.h (hunk 1 above)
Comment (jl-wynen) [reply dataset/include/scipp/dataset/dataset_util.h:36]: Not possible: 
```.py
---------------------------------------------------------------------------
DTypeError                                Traceback (most recent call last)
<ipython-input-14-3a82f34c1694> in <module>
----> 1 da.bins.coords['abc'] = var

DTypeError: Expected item dtype VariableView, got DataArray.
```

File: dataset/test/data_array_test.cpp (hunk 4 above)
Comment (jl-wynen) [reply dataset/test/data_array_test.cpp:153]: This is fine. `sc.scalar(da)` makes a shallow copy of `da`. So the new coord does not have 'abc' as a coord itself.

File: variable/variable.cpp [hunk 5]
```diff
@@ -29,6 +27,45 @@ Variable::Variable(const Dimensions &dims, VariableConceptHandle data)
...
     : Variable(m.value() * units::Unit(m.units())) {}
 
+namespace {
+void check_nested_in_assign(const Variable &lhs, const Variable &rhs) {
+  if (!rhs.is_valid() || rhs.dtype() != dtype<Variable>) {
+    return;
+  }
+  // In principle we should also check when the RHS contains DataArrays or
+  // Datasets. But those are copied when stored in Variables,
+  // so no check needed here.
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply variable/variable.cpp:37]: It is enough in most cases. You need to nest this deep enough for the shallow copy to not handle. Or do things like
```.py
v = sc.scalar(da)
v.value.attrs['abc'] = v
```

File: dataset/test/data_array_test.cpp (hunk 4 above)
Comment (SimonHeybrock) [reply dataset/test/data_array_test.cpp:153]: Is it fine though? What happens now if you `del da`? Doesn't it leak?

Comment (jl-wynen) [reply dataset/test/data_array_test.cpp:153]: Why would it?
I can this through valgrind with and without `del` and it shows no difference. Also, there is no difference whether I add the coord or not. Which makes sense, I think,

Comment (SimonHeybrock) [reply dataset/test/data_array_test.cpp:153]: Right, I think you'd need a more contrieved example.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**